File: Android App/app/src/main/java/com/fica/events/data/api/ChatWebSocket.kt
package com.fica.events.data.api

import com.fica.events.data.models.Message
import com.google.gson.Gson
import com.google.gson.JsonParser
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener

object ChatWebSocket {

    // Production WebSocket — wss://eventsfiji.cloud/ws (Nginx proxies to backend on :5000)
    // For local dev against the emulator, swap to "ws://10.0.2.2:5000/ws".
    private const val WS_URL = "wss://eventsfiji.cloud/ws"

    private var webSocket: WebSocket? = null
    private var userId: Int? = null
    // Cached JWT so an onFailure-triggered reconnect can re-auth with the
    // server's upgrade-URL check. Cleared on disconnect().
    private var authToken: String? = null
    private val gson = Gson()
    private val client = OkHttpClient()

    // Handlers keyed by conversation pair "min-max"
    private val conversationHandlers = mutableMapOf<String, (Message) -> Unit>()

    // Global handler for chat list refresh
    var onAnyMessage: ((Message) -> Unit)? = null

    // Panel discussion open/close flips, broadcast from the admin UI.
    // (sessionId, discussionEnabled) — tokenised so the list and a
    // currently-open detail screen can subscribe independently.
    private val panelDiscussionHandlers = mutableMapOf<java.util.UUID, (Int, Boolean) -> Unit>()
    // Voting open/closed + results-visibility flips.
    private val votingOpenHandlers = mutableMapOf<java.util.UUID, (Boolean) -> Unit>()
    private val votingResultsHandlers = mutableMapOf<java.util.UUID, (Boolean) -> Unit>()

    /**
     * Open an authenticated WebSocket. The JWT is appended to the upgrade
     * URL as `?token=<jwt>`; the server verifies it on upgrade and rejects
     * missing/invalid tokens with HTTP 401. The userId argument is kept
     * for caller ergonomics but identity is now derived from the verified
     * token on the server.
     */
    fun connect(userId: Int, token: String) {
        if (webSocket != null) return
        this.userId = userId
        this.authToken = token

        val encodedToken = java.net.URLEncoder.encode(token, "UTF-8")
        val request = Request.Builder().url("$WS_URL?token=$encodedToken").build()
        webSocket = client.newWebSocket(request, object : WebSocketListener() {
            override fun onOpen(ws: WebSocket, response: Response) {
                // Server no longer trusts this message; sent only so older
                // backends during a rolling upgrade still see a join.
                ws.send("""{"event":"join","userId":$userId}""")
            }

            override fun onMessage(ws: WebSocket, text: String) {
                handleMessage(text)
            }

            override fun onFailure(ws: WebSocket, t: Throwable, response: Response?) {
                webSocket = null
                // 401 on the upgrade = the token is stale (server rotated
                // JWT_SECRET, or the token expired). Boot to login rather
                // than looping forever — that's what made the app look
                // frozen after a backend deploy. AuthManager.logout()
                // notifies the root NavHost via the logout-listener API,
                // which pops back to Login.
                if (response?.code == 401) {
                    com.fica.events.data.auth.AuthManager.logout()
                    return
                }
                // Transient network failure — retry after 3s.
                Thread.sleep(3000)
                val uid = ChatWebSocket.userId
                val tkn = ChatWebSocket.authToken
                if (uid != null && tkn != null) connect(uid, tkn)
            }

            override fun onClosed(ws: WebSocket, code: Int, reason: String) {
                webSocket = null
            }
        })
    }

    fun disconnect() {
        webSocket?.close(1000, "logout")
        webSocket = null
        authToken = null
        conversationHandlers.clear()
        onAnyMessage = null
        panelDiscussionHandlers.clear()
        votingOpenHandlers.clear()
        votingResultsHandlers.clear()
    }

    fun addConversationHandler(myId: Int, otherId: Int, handler: (Message) -> Unit) {
        val key = "${minOf(myId, otherId)}-${maxOf(myId, otherId)}"
        conversationHandlers[key] = handler
    }

    fun removeConversationHandler(myId: Int, otherId: Int) {
        val key = "${minOf(myId, otherId)}-${maxOf(myId, otherId)}"
        conversationHandlers.remove(key)
    }

    /** Subscribe to live panel discussion open/close events. Keep the
     *  returned token and pass it to [removePanelDiscussionHandler] on
     *  dispose so the singleton doesn't leak stale observers. */
    fun addPanelDiscussionHandler(handler: (sessionId: Int, enabled: Boolean) -> Unit): java.util.UUID {
        val id = java.util.UUID.randomUUID()
        panelDiscussionHandlers[id] = handler
        return id
    }

    fun removePanelDiscussionHandler(id: java.util.UUID) {
        panelDiscussionHandlers.remove(id)
    }

    fun addVotingOpenHandler(handler: (Boolean) -> Unit): java.util.UUID {
        val id = java.util.UUID.randomUUID()
        votingOpenHandlers[id] = handler
        return id
    }

    fun removeVotingOpenHandler(id: java.util.UUID) { votingOpenHandlers.remove(id) }

    fun addVotingResultsHandler(handler: (Boolean) -> Unit): java.util.UUID {
        val id = java.util.UUID.randomUUID()
        votingResultsHandlers[id] = handler
        return id
    }

    fun removeVotingResultsHandler(id: java.util.UUID) { votingResultsHandlers.remove(id) }

    private fun handleMessage(text: String) {
        try {
            val json = JsonParser.parseString(text).asJsonObject
            val event = json.get("event")?.asString ?: return
            when (event) {
                "new_message" -> {
                    val message = gson.fromJson(json.get("data"), Message::class.java)
                    val key = "${minOf(message.sender_id, message.receiver_id)}-${maxOf(message.sender_id, message.receiver_id)}"
                    conversationHandlers[key]?.invoke(message)
                    onAnyMessage?.invoke(message)
                }
                "panel_discussion_changed" -> {
                    val data = json.get("data")?.asJsonObject ?: return
                    val sessionId = data.get("session_id")?.asInt ?: return
                    val enabled = data.get("discussion_enabled")?.asBoolean ?: return
                    // Copy to avoid ConcurrentModification if a handler
                    // unregisters itself during iteration.
                    panelDiscussionHandlers.values.toList().forEach { it(sessionId, enabled) }
                }
                "voting_open_changed" -> {
                    val data = json.get("data")?.asJsonObject ?: return
                    val open = data.get("voting_open")?.asBoolean ?: return
                    votingOpenHandlers.values.toList().forEach { it(open) }
                }
                "voting_results_visibility_changed" -> {
                    val data = json.get("data")?.asJsonObject ?: return
                    val visible = data.get("voting_results_visible")?.asBoolean ?: return
                    votingResultsHandlers.values.toList().forEach { it(visible) }
                }
            }
        } catch (_: Exception) {}
    }
}

File: IOS App/Events/Services/ChatWebSocket.swift
import Foundation

@MainActor
@Observable
final class ChatWebSocket {
    static let shared = ChatWebSocket()

    private var task: URLSessionWebSocketTask?
    private var isConnected = false
    private var userId: Int?
    // Cached JWT so a failure-triggered reconnect has what it needs to
    // re-auth on the upgrade URL. Cleared on disconnect().
    private var token: String?

    // Callbacks keyed by conversation pair "min-max"
    private var messageHandlers: [String: (Message) -> Void] = [:]
    // Global handler for chat list updates
    var onAnyMessage: ((Message) -> Void)?
    // Panel discussion open/close flips, broadcast from the admin UI.
    // (sessionId, discussionEnabled) — multiple subscribers supported so
    // the Panels list and a currently-open Panel Detail can both react.
    private var panelDiscussionHandlers: [UUID: (Int, Bool) -> Void] = [:]
    // Voting open/closed flips (admin → all delegates).
    private var votingOpenHandlers: [UUID: (Bool) -> Void] = [:]
    // Voting results visibility flips (admin → all delegates).
    private var votingResultsHandlers: [UUID: (Bool) -> Void] = [:]

    // Production WebSocket — wss://eventsfiji.cloud/ws (Nginx proxies to backend on :5000)
    // For local dev, flip to "ws://localhost:5000/ws" (simulator) or your LAN IP (device).
    private let wsURL = "wss://eventsfiji.cloud/ws"

    /// Open an authenticated WebSocket. The JWT is appended as a query
    /// param; the server verifies it on upgrade and rejects invalid /
    /// missing tokens with HTTP 401. The userId argument is kept for
    /// backward-compat / caller ergonomics but is no longer trusted by
    /// the server — identity is derived from the verified token.
    func connect(userId: Int, token: String) {
        guard !isConnected else { return }
        self.userId = userId
        self.token = token

        // URL-encode the token just in case a future token format includes
        // `+`, `/`, or `=` (base64url JWTs don't, but defensive).
        let allowed = CharacterSet.urlQueryAllowed
        let encoded = token.addingPercentEncoding(withAllowedCharacters: allowed) ?? token
        guard let url = URL(string: "\(wsURL)?token=\(encoded)") else { return }
        task = URLSession.shared.webSocketTask(with: url)
        task?.resume()
        isConnected = true

        // The server no longer reads userId from the join message, but we
        // keep sending it so a rolling upgrade where some backends still
        // expect the message won't regress. Server-side it's a no-op.
        let joinMsg = #"{"event":"join","userId":\#(userId)}"#
        task?.send(.string(joinMsg)) { _ in }

        listen()
    }

    func disconnect() {
        task?.cancel(with: .normalClosure, reason: nil)
        task = nil
        isConnected = false
        token = nil
        messageHandlers.removeAll()
        onAnyMessage = nil
        panelDiscussionHandlers.removeAll()
        votingOpenHandlers.removeAll()
        votingResultsHandlers.removeAll()
    }

    func addConversationHandler(myId: Int, otherId: Int, handler: @escaping (Message) -> Void) {
        let key = "\(min(myId, otherId))-\(max(myId, otherId))"
        messageHandlers[key] = handler
    }

    func removeConversationHandler(myId: Int, otherId: Int) {
        let key = "\(min(myId, otherId))-\(max(myId, otherId))"
        messageHandlers.removeValue(forKey: key)
    }

    /// Subscribe to panel discussion open/close push events. Returns a
    /// token; pass it back to `removePanelDiscussionHandler(_:)` on
    /// disappear to avoid leaking observers across view instances.
    @discardableResult
    func addPanelDiscussionHandler(_ handler: @escaping (_ sessionId: Int, _ enabled: Bool) -> Void) -> UUID {
        let id = UUID()
        panelDiscussionHandlers[id] = handler
        return id
    }

    func removePanelDiscussionHandler(_ id: UUID) {
        panelDiscussionHandlers.removeValue(forKey: id)
    }

    @discardableResult
    func addVotingOpenHandler(_ handler: @escaping (_ open: Bool) -> Void) -> UUID {
        let id = UUID()
        votingOpenHandlers[id] = handler
        return id
    }

    func removeVotingOpenHandler(_ id: UUID) {
        votingOpenHandlers.removeValue(forKey: id)
    }

    @discardableResult
    func addVotingResultsHandler(_ handler: @escaping (_ visible: Bool) -> Void) -> UUID {
        let id = UUID()
        votingResultsHandlers[id] = handler
        return id
    }

    func removeVotingResultsHandler(_ id: UUID) {
        votingResultsHandlers.removeValue(forKey: id)
    }

    private func listen() {
        task?.receive { [weak self] result in
            switch result {
            case .success(let wsMessage):
                if case .string(let text) = wsMessage {
                    self?.handleRaw(text)
                }
                // Keep listening
                self?.listen()
            case .failure:
                Task { @MainActor in
                    self?.isConnected = false
                    // Reconnect after 3 seconds using the cached token. If
                    // the token has expired the server will 401 on upgrade
                    // and the next receive() will fail — loop will back off
                    // until the user next logs in.
                    try? await Task.sleep(for: .seconds(3))
                    if let uid = self?.userId, let tkn = self?.token {
                        self?.connect(userId: uid, token: tkn)
                    }
                }
            }
        }
    }

    private func handleRaw(_ text: String) {
        guard let data = text.data(using: .utf8) else { return }

        // Peek at event type first so we can dispatch to the right decoder.
        struct EventPeek: Decodable { let event: String }
        guard let peek = try? JSONDecoder().decode(EventPeek.self, from: data) else { return }

        switch peek.event {
        case "new_message":
            struct MessagePayload: Decodable { let event: String; let data: Message }
            guard let payload = try? JSONDecoder().decode(MessagePayload.self, from: data) else { return }
            let message = payload.data
            Task { @MainActor [weak self] in
                let key = "\(min(message.sender_id, message.receiver_id))-\(max(message.sender_id, message.receiver_id))"
                self?.messageHandlers[key]?(message)
                self?.onAnyMessage?(message)
            }

        case "panel_discussion_changed":
            struct PanelDiscussionData: Decodable {
                let session_id: Int
                let discussion_enabled: Bool
            }
            struct PanelPayload: Decodable { let event: String; let data: PanelDiscussionData }
            guard let payload = try? JSONDecoder().decode(PanelPayload.self, from: data) else { return }
            let sessionId = payload.data.session_id
            let enabled = payload.data.discussion_enabled
            Task { @MainActor [weak self] in
                guard let self else { return }
                for handler in self.panelDiscussionHandlers.values {
                    handler(sessionId, enabled)
                }
            }

        case "voting_open_changed":
            struct VotingOpenData: Decodable { let voting_open: Bool }
            struct VotingOpenPayload: Decodable { let event: String; let data: VotingOpenData }
            guard let payload = try? JSONDecoder().decode(VotingOpenPayload.self, from: data) else { return }
            let open = payload.data.voting_open
            Task { @MainActor [weak self] in
                guard let self else { return }
                for handler in self.votingOpenHandlers.values { handler(open) }
            }

        case "voting_results_visibility_changed":
            struct ResultsData: Decodable { let voting_results_visible: Bool }
            struct ResultsPayload: Decodable { let event: String; let data: ResultsData }
            guard let payload = try? JSONDecoder().decode(ResultsPayload.self, from: data) else { return }
            let visible = payload.data.voting_results_visible
            Task { @MainActor [weak self] in
                guard let self else { return }
                for handler in self.votingResultsHandlers.values { handler(visible) }
            }

        default:
            break
        }
    }
}
